File: haskell/test/DeserializeSpec.hs
module DeserializeSpec (spec) where

import Data.Aeson (eitherDecode)
import Data.Either (isLeft)
import qualified Data.ByteString.Lazy.Char8 as LBS
import Test.Hspec
import TestTypes

-- Interop JSON — identical strings used across all 12 language libraries
circleJson, rectangleJson, carJson, truckJson :: String
circleJson = "{\"#type\":\"Circle\",\"radius\":4}"
rectangleJson = "{\"#type\":\"Rectangle\",\"width\":10,\"height\":5}"
carJson = "{\"#type\":\"Car\",\"make\":\"Toyota\"}"
truckJson = "{\"#type\":\"Truck\",\"payloadTons\":5}"

spec :: Spec
spec = describe "deserialize" $ do

  it "decode_whenDeserializingCircle_resolvesCorrectType" $ do
    -- Act
    let result = eitherDecode (LBS.pack circleJson) :: Either String Shape
    -- Assert
    result `shouldBe` Right (Circle { radius = 4.0 })

  it "decode_whenDeserializingRectangle_resolvesCorrectType" $ do
    -- Act
    let result = eitherDecode (LBS.pack rectangleJson) :: Either String Shape
    -- Assert
    result `shouldBe` Right (Rectangle { width = 10.0, height = 5.0 })

  it "decode_whenDeserializingCar_resolvesCorrectType" $ do
    -- Act
    let result = eitherDecode (LBS.pack carJson) :: Either String Vehicle
    -- Assert
    result `shouldBe` Right (PersonalVehicle { make = "Toyota" })

  it "decode_whenDeserializingTruck_resolvesCorrectType" $ do
    -- Act
    let result = eitherDecode (LBS.pack truckJson) :: Either String Vehicle
    -- Assert
    result `shouldBe` Right (TransportationVehicle { payloadTons = 5.0 })

  it "decode_whenHashTypeIsNotFirstProperty_stillDeserializes" $ do
    -- Act
    let result = eitherDecode (LBS.pack "{\"radius\":4,\"#type\":\"Circle\"}") :: Either String Shape
    -- Assert
    result `shouldBe` Right (Circle { radius = 4.0 })

  it "decode_whenUnknownTypeName_failsToDeserialize" $ do
    -- Act
    let result = eitherDecode (LBS.pack "{\"#type\":\"Hexagon\",\"sides\":6}") :: Either String Shape
    -- Assert
    result `shouldSatisfy` isLeft

File: scala/src/test/scala/jsontag/DeserializeSpec.scala
package jsontag

import munit.FunSuite

class DeserializeSpec extends FunSuite:

  // Interop JSON — identical strings used across all 12 language libraries
  val circleJson = """{"#type":"Circle","radius":4}"""
  val rectangleJson = """{"#type":"Rectangle","width":10,"height":5}"""
  val carJson = """{"#type":"Car","make":"Toyota"}"""
  val truckJson = """{"#type":"Truck","payloadTons":5}"""

  test("deserialize, when deserializing Circle, resolves correct type"):
    // Act
    val shape = JsonTag.read[Shape](circleJson)

    // Assert
    assertEquals(shape, Circle(4.0))

  test("deserialize, when deserializing Rectangle, resolves correct type"):
    // Act
    val shape = JsonTag.read[Shape](rectangleJson)

    // Assert
    assertEquals(shape, Rectangle(10.0, 5.0))

  test("deserialize, when deserializing Car, resolves correct type"):
    // Act
    val vehicle = JsonTag.read[Vehicle](carJson)

    // Assert
    assertEquals(vehicle, PersonalVehicle("Toyota"))

  test("deserialize, when deserializing Truck, resolves correct type"):
    // Act
    val vehicle = JsonTag.read[Vehicle](truckJson)

    // Assert
    assertEquals(vehicle, TransportationVehicle(5.0))

  test("deserialize, when #type is not first property, still deserializes"):
    // Act
    val shape = JsonTag.read[Shape]("""{"radius":4,"#type":"Circle"}""")

    // Assert
    assertEquals(shape, Circle(4.0))

  test("deserialize, when unknown type name, fails to deserialize"):
    // Act & Assert
    intercept[Exception]:
      JsonTag.read[Shape]("""{"#type":"Hexagon","sides":6}""")
